Refactor AliasOrIndex abstraction.

martijnvg opened the pull request:
In order to prepare the `AliasOrIndex` abstraction for the introduction of data streams,
the abstraction needs to be made more flexible, because currently it really can be only
an alias or an index.

The following changes were made:
* Renamed `AliasOrIndex` to `IndexAbstraction`.
* Introduced a `IndexSpace.Type` enum to indicate what a `IndexSpace` instance is.
* Replaced the `isAlias()` method that returns a boolean with the `getType()` method that returns the new Type enum.
* Moved `getWriteIndex()` up from the `IndexSpace.Alias` to the `IndexSpace` interface.
* Moved `getAliasName()` up from the `IndexSpace.Alias` to the `IndexSpace` interface and renamed it to `getName()`.
* Removed unnecessary casting to `IndexSpace.Alias` by just checking the `getType()` method.

Relates to #53100

martijnvg committed:
Refactor AliasOrIndex abstraction.

In order to prepare the `AliasOrIndex` abstraction for the introduction of data streams,
the abstraction needs to be made more flexible, because currently it really can be only
an alias or an index.

* Introduced a `AliasOrIndex.Type` enum to indicate what a `AliasOrIndex` instance is.
* Replaced the `isAlias()`` method that returns a boolean with the `getType()`` method that returns the new Type enum.
* Moved `getWriteIndex()`` up from the `AliasOrIndex.Alias` to the `AliasOrIndex` interface.
* Moved `g`etAliasName()`` up from the `AliasOrIndex.Alias` to the `AliasOrIndex` interface and renamed it to `getName()``.
* Removed unnecessary casting to `AliasOrIndex.Alias` by just checking the `getType()`` method.

Finally `AliasOrIndex` should be renamed to reflect that it can be more than just an index or alias, since
in the near future it can also be a data stream. The name AliasOrIndexOrDataStream is not appealing to me.
We can rename it to `Namespace`, but that sounds to generic to me. `ResolvedIndicesExpression` sounds better
to me, since it reflects more what it is (an expression from api that has been resolved to alias/index/datasteam),
but the name itself is a bit on the long side.

Relates to #53100

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

martijnvg committed:
fixed checkstyle violation

martijnvg committed:
tweaked logic

martijnvg committed:
Merge remote-tracking branch 'es/master' into improve_alias_or_index_abstraction

martijnvg committed:
fixed compile error after merging in master

martijnvg committed:
renamed AliasOrIndex to IndexSpace and added jdocs

martijnvg committed:
Merge remote-tracking branch 'es/master' into improve_alias_or_index_abstraction

danhermann commented on server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -263,16 +263,15 @@ static String generateRolloverIndexName(String sourceIndexName, IndexNameExpress
         }
-        if (aliasOrIndex.isAlias() == false) {
+        if (indexSpace.getType() != IndexSpace.Type.ALIAS) {
             throw new IllegalArgumentException("source alias is a concrete index");
```
Looks like a lot of error messages such as this one that assume only alias or index types would benefit from wording changes that accommodate data streams, too.

martijnvg replied:
Good point. I will go over these checks and messages and adjust.

danhermann commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -281,18 +281,16 @@ static boolean resolvePipelines(final DocWriteRequest<?> originalRequest, final
-                if (indexOrAlias != null && indexOrAlias.isAlias()) {
-                    AliasOrIndex.Alias alias = (AliasOrIndex.Alias) indexOrAlias;
-                    indexMetaData = alias.getWriteIndex();
+                IndexSpace indexOrAlias = metaData.getAliasAndIndexLookup().get(indexRequest.index());
```
We'll probably want to rename that variable from `indexOrAlias`. Can be done in a later PR, though.

danhermann commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexSpace.java:
```diff
@@ -35,30 +34,68 @@
-     * @return whether this alias/index is hidden or not
+     * A write index is a dedicated concrete index, that accepts all the new documents that belong to an index space.
+     *
+     * A write index may also be a regular concrete index of a index space and may be therefor also be returned
```
nit: "may be therefor also be returned" -> "may therefore also be returned"

danhermann commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java:
```diff
@@ -14,7 +14,7 @@
 import org.elasticsearch.action.support.IndicesOptions;
 import org.elasticsearch.cluster.metadata.AliasMetaData;
-import org.elasticsearch.cluster.metadata.AliasOrIndex;
+import org.elasticsearch.cluster.metadata.IndexSpace;
```
Will we want to eventually rename this class to something like `IndexSpaceResolver`?

martijnvg replied:
Yes, I think so.

danhermann reviewed (approved):
LGTM. I left a few comments that are minor, not actionable, and/or addressable in later PRs.

martijnvg committed:
Merge remote-tracking branch 'es/master' into improve_alias_or_index_abstraction

martijnvg committed:
applied review comments and
renamed `MetaData#aliasAndIndexLookup` field (and getter) to `MetaData#indexSpaceLookup`.

henningandersen commented:
I wonder if we should use `AbstractIndex` instead of `IndexSpace`? It could be good to come up with a terminology for alias/index/datastream that we would consider using in documentation too.

danhermann commented:
> I wonder if we should use `AbstractIndex` instead of `IndexSpace`? It could be good to come up with a terminology for alias/index/datastream that we would consider using in documentation too.

We considered a bunch of different options including `AliasOrIndexOrDataStream` (not really),  `Namespace`, `ResolvedIndexExpression`, and `IndexFamily` and landed on `IndexSpace` as the best of them. I agree that it would be good to have one that works well in documentation, too. Naming things is hard.

martijnvg commented:
Dan and I did came up with the following alternative: `IndexAbstraction`.

`AbstractIndex` is very 'programming' term and could indicate to someone reading that name, that it isn't concrete. Whereas `IndexAbstraction` reads as an abstraction of an index, which represents something and which could be a concrete index, an alias or a data stream.

martijnvg committed:
Merge remote-tracking branch 'es/master' into improve_alias_or_index_abstraction

martijnvg committed:
changed MetDataRolloverService to use IndexSpace after merge
and fixed test

martijnvg committed:
renamed IndexSpace to IndexAbstraction

martijnvg committed:
tweaked checks to also work if more index abstraction types are introduced.

martijnvg committed:
List.of(...) everywhere

henningandersen commented on server/src/test/java/org/elasticsearch/action/admin/indices/rollover/MetaDataRolloverServiceTests.java:
```diff
@@ -212,7 +212,7 @@ public void testValidation() {
         exception = expectThrows(IllegalArgumentException.class, () ->
             MetaDataRolloverService.validate(metaData, randomFrom(index1, index2)));
-        assertThat(exception.getMessage(), equalTo("source alias is a concrete index"));
+        assertThat(exception.getMessage(), equalTo("source alias is not an alias"));
```
I think the change to the exception message is not really a big issue. But with data streams also appearing, it could be nice to stick to the old message since that would give you some information on whether it is a datastream or concrete index. I think we could easily generate such a message by adding a display name ("concrete index") to the IndexAbstract.Type enum?

Also, this does make me wonder if `Type.INDEX` should be `Type.CONCRETE_INDEX` for consistency with that display name? Not truly important to me, feel free to ignore this part.

henningandersen commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/SamlServiceProviderIndex.java:
```diff
@@ -187,11 +187,11 @@ public void close() {
-        } else if (aliasInfo.isAlias() == false) {
-            logger.warn("service provider index [{}] exists as a concrete index, but it should be an alias", ALIAS_NAME);
+        } else if (aliasInfo.getType() != IndexAbstraction.Type.ALIAS) {
+            logger.warn("service provider index [{}] does not exist as an alias, but it should be", ALIAS_NAME);
```
Same comment on retaining the original message here.

henningandersen commented on x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/integration/MlConfigMigratorIT.java:
```diff
@@ -83,8 +83,8 @@ public void setUpTests() {
         MetaData metaData = mock(MetaData.class);
-        SortedMap<String, AliasOrIndex> aliasOrIndexSortedMap = new TreeMap<>();
-        when(metaData.getAliasAndIndexLookup()).thenReturn(aliasOrIndexSortedMap);
+        SortedMap<String, IndexAbstraction> indexSpaceSortedMap = new TreeMap<>();
```
Remove "Space" from variable name?

henningandersen commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java:
```diff
@@ -228,12 +228,12 @@ static String getPutMappingIndexOrAlias(PutMappingRequest request, List<String>
-        } else if (aliasOrIndex.isAlias()) {
-            throw new IllegalStateException("concrete index [" + concreteIndexName + "] is an alias but should not be");
+        } else if (indexAbstraction.getType() != IndexAbstraction.Type.INDEX) {
+            throw new IllegalStateException("concrete index [" + concreteIndexName + "] is not an index but should be");
```
I think we should also include the display name of the actual type here.

henningandersen reviewed (approved):
LGTM.

I left a few smaller comments to consider.

martijnvg committed:
Merge remote-tracking branch 'es/master' into improve_alias_or_index_abstraction

martijnvg committed:
iter

martijnvg committed:
fixed checkstyle violation

martijnvg committed:
fixed checkstyle violation (2)
